Emit FLL from wrangler transform

rohitsinha54 opened the pull request:
- Emit basic FLL from Wrangler Transform
- See details [here](https://issues.cask.co/browse/CDAP-14005?focusedCommentId=33734&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-33734)

![image](https://user-images.githubusercontent.com/1472867/45514369-05f9d980-b75a-11e8-8ac0-95d2dddac89c.png)

[Issue](https://issues.cask.co/browse/CDAP-14005)

rohitsinha54 committed:
Emit FLL from wrangler transform

albertshau commented on wrangler-transform/src/main/java/co/cask/wrangler/Wrangler.java:
```diff
@@ -56,6 +59,7 @@
 import co.cask.wrangler.registry.SystemDirectiveRegistry;
 import co.cask.wrangler.registry.UserDirectiveRegistry;
 import com.google.common.collect.ImmutableMap;
+import edu.emory.mathcs.backport.java.util.Collections;
```
wrong Collections

albertshau commented on wrangler-transform/src/main/java/co/cask/wrangler/Wrangler.java:
```diff
@@ -233,6 +239,38 @@ public void configurePipeline(PipelineConfigurer configurer) throws IllegalArgum
+
+  private boolean checkSchema(Schema schema, String name) {
+    if (schema == null) {
+      LOG.debug(String.format("The %s schema is null.", name));
```
Log messages should tell somebody reading it everything they need to know without looking at any context. Better to make it:

is null. Field level lineage will not be recorded.

albertshau commented on wrangler-transform/src/main/java/co/cask/wrangler/Wrangler.java:
```diff
@@ -233,6 +239,38 @@ public void configurePipeline(PipelineConfigurer configurer) throws IllegalArgum
+      return false;
+    }
+    if (schema.getFields() == null) {
+      LOG.debug(String.format("The %s schema fields are null.", name));
```
is null. Field level lineage will not be recorded.

albertshau commented on wrangler-transform/src/main/java/co/cask/wrangler/Wrangler.java:
```diff
@@ -233,6 +239,38 @@ public void configurePipeline(PipelineConfigurer configurer) throws IllegalArgum
+      //noinspection ConstantConditions
+      outputFields = outputSchema.getFields().stream().map(Schema.Field::getName).collect(Collectors.toList());
+    }
+    FieldOperation dataPrepOperation = new FieldTransformOperation("DataPrep", config.directives, inputFields,
```
The other names seem to be verbs, whereas this is a noun. I would call it 'Prepare' or 'Prepare Data' (are spaces allowed?)

albertshau reviewed (commented):
from the screenshot I don't see any separators between the directives, which makes it hard to read. Are they stripped somehow?

elfenheart commented:
Would it be possible to join the list of operations with some kind of a delimiter? Perhaps a `;`? The description is not really readable for wrangler. @bdmogal any input?

rohitsinha54 commented:
@albertshau Addressed comments. 
The separators are added in V2 format I believe so I have to migrate the directive.
![image](https://user-images.githubusercontent.com/1472867/45515789-0a27f600-b75e-11e8-8781-fcd7bcc47bff.png)

rohitsinha54 commented:
@elfenheart Please have another look. Added delimiters.

albertshau commented on wrangler-transform/src/main/java/co/cask/wrangler/Wrangler.java:
```diff
@@ -233,6 +239,40 @@ public void configurePipeline(PipelineConfigurer configurer) throws IllegalArgum
+      outputFields = outputSchema.getFields().stream().map(Schema.Field::getName).collect(Collectors.toList());
+    }
+    FieldOperation dataPrepOperation = new FieldTransformOperation("Prepare Data",
+                                                                   new MigrateToV2(config.directives).migrate(),
```
I assume this doesn't cause problems if the directives are already in "v2" format? What format is "v1"? It couldn't have been splitting on spaces.

IMO, newlines are the most readable.

rohitsinha54 replied:
There is MigrateToV2Test which tests that new format can be passed in so it can handle that.
It does not split on whitespace it sees the directive name and then parses the required thing so it has logic for each and every directive 

https://github.com/data-integrations/wrangler/blob/develop/wrangler-core/src/main/java/co/cask/wrangler/parser/MigrateToV2.java#L67

rohitsinha54 replied:
The directives are separated by new line but the UI does not show it.

@elfenheart Will work on fixing that.

albertshau reviewed (approved):
lgtm
